evidence/pool.go

```diff
@@ -138,29 +134,185 @@ func (evpool *Pool) AddEvidence(ev types.Evidence) error {
 	return nil
 }
 
-// Verify verifies the evidence against the node's (or evidence pool's) state. More specifically, to validate
-// evidence against state is to validate it against the nodes own header and validator set for that height. This ensures
-// as well as meeting the evidence's own validation rules, that the evidence hasn't expired, that the validator is still
-// bonded and that the evidence can be committed to the chain.
-func (evpool *Pool) Verify(evidence types.Evidence) error {
-	if evpool.IsCommitted(evidence) {
-		return errors.New("evidence was already committed")
+// AddEvidenceFromConsensus should be exposed only to the consensus so it can add evidence to the pool
+// directly without the need for verification.
+func (evpool *Pool) AddEvidenceFromConsensus(ev types.Evidence, time time.Time, valSet *types.ValidatorSet) error {
+	var (
+		vals       []*types.Validator
+		totalPower int64
+	)
+
+	if evpool.isPending(ev) {
+		return fmt.Errorf("evidence already verified and added, %s", ev.String()) // we already have this evidence
 	}
-	// We have already verified this piece of evidence - no need to do it again
-	if evpool.IsPending(evidence) {
-		return nil
+
+	switch ev := ev.(type) {
+	case *types.DuplicateVoteEvidence:
+		_, val := valSet.GetByAddress(ev.VoteA.ValidatorAddress)
+		vals = append(vals, val)
+		totalPower = valSet.TotalVotingPower()
+	default:
+		return fmt.Errorf("unrecognized evidence: %v", ev)
+	}
+
+	evInfo := &Info{
+		Evidence:         ev,
+		Time:             time,
+		Validators:       vals,
+		TotalVotingPower: totalPower,
+	}
+
+	if err := evpool.addPendingEvidence(evInfo); err != nil {
+		return fmt.Errorf("database error when adding evidence from consensus: %w", err)
+	}
+
+	evpool.evidenceList.PushBack(ev)
+
+	evpool.logger.Info("Verified new evidence of byzantine behavior", "evidence", ev)
+
+	return nil
+}
+
+// CheckEvidence takes an array of evidence from a block and verifies all the evidence there.
+// If it has already verified the evidence then it jumps to the next one. It ensures that no
+// evidence has already been committed or is being proposed twice. It also adds any
+// evidence that it doesn't currently have so that it can quickly form ABCI Evidence later.
+// Check evidence returns the hash of the entire evidence list so that it can be compares with
+// the headers EvidenceHash.
+func (evpool *Pool) CheckEvidence(evList types.EvidenceList) error {
+	hashes := make([][]byte, len(evList))
+	for idx, ev := range evList {
+
+		ok := evpool.fastCheck(ev)
```

Comment:
```suggestion
		ok := evpool.fastVerify(ev)
```

? is there a difference between check and verify?

Reply:
Well I called it check instead of verify because we aren't really verifying it, just checking that we verified the evidence before. And if we have and it's light client evidence, we make sure that the validators we found were malicious have signed in the commit (as the hashes of the light client evidence are identical even when different sets of validators have signed the commit)